### PR TITLE
Spring Boot Security 1.0

### PR DESCRIPTION
**Pull request by aksoli666**
added spring-boot-starter-security
added SecurityConfig
created CustomUserDetailsService
implemented UserDetails in the User
/auth/** -> public, other -> accessible before auth added role entity and repo layer
added all liquibase changesets
used @PreAuthorisation

### CODE REVIEW

**Comment by aksoli666** [comment]
Hi, I have a problem sending requests to create a book from a user with the role of user, the request is successful, but it should not work like this. How do I properly fix the code because I can't find the error?

File: src/main/java/shop/dto/request/UserRegistrationRequestDto.java [hunk 1]
```diff
@@ -26,4 +28,5 @@ public class UserRegistrationRequestDto {
     @NotBlank
     private String lastName;
     private String shippingAddress;
+    private Set<Role> roles;
```
Comment (liliia-ponomarenko) [inline src/main/java/shop/dto/request/UserRegistrationRequestDto.java]: NotEmpty

File: src/main/java/shop/entity/User.java [hunk 2]
```diff
@@ -26,4 +39,43 @@ public class User {
     @Column(nullable = false)
     private String lastName;
     private String shippingAddress;
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (liliia-ponomarenko) [inline src/main/java/shop/entity/User.java]: Don't use EAGER fetch type

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml [hunk 3]
```diff
@@ -41,7 +41,7 @@ databaseChangeLog:
                   type: varchar(255)
               - column:
                   name: is_deleted
-                  type: boolean
+                  type: bit
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml]:
```suggestion
                  type: tinyint(1)
```

File: src/main/resources/db/changelog/changes/04-create-users-table.yaml [hunk 4]
```diff
@@ -37,4 +37,9 @@ databaseChangeLog:
               - column:
                   name: shipping_address
                   type: varchar(255)
-
+              - column:
+                  name: is_deleted
+                  type: bit
+                  constraints:
+                    nullable: false
+
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/04-create-users-table.yaml]: This is normal at the training stage, but you should understand that on real projects you will never change existing files, but always create a new one for new changes

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s fix a few mistakes ;)

File: src/main/java/shop/config/SecurityConfig.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
+import org.springframework.security.config.http.SessionCreationPolicy;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+import shop.service.user.details.CustomUserDetailsService;
+
+@EnableWebSecurity
+@EnableMethodSecurity
```
Comment (Elena-Bruyako) [inline src/main/java/shop/config/SecurityConfig.java:16]: leave only one annotation

File: src/main/java/shop/dto/request/CreateBookRequestDto.java [hunk 6]
```diff
@@ -21,4 +21,5 @@ public class CreateBookRequestDto {
     private BigDecimal price;
     private String description;
     private String coverImage;
+
```
Comment (Elena-Bruyako) [inline src/main/java/shop/dto/request/CreateBookRequestDto.java]: remove empty line

File: src/main/java/shop/dto/request/UserRegistrationRequestDto.java [hunk 7]
```diff
@@ -26,4 +29,6 @@ public class UserRegistrationRequestDto {
     @NotBlank
     private String lastName;
     private String shippingAddress;
+    @NotEmpty
+    private Set<Role> roles;
```
Comment (Elena-Bruyako) [inline src/main/java/shop/dto/request/UserRegistrationRequestDto.java]: don't need to provide roles in UserRegistrationRequestDto
```suggestion
```

File: src/main/java/shop/dto/responce/UserResponseDto.java [hunk 8]
```diff
@@ -11,4 +13,5 @@ public class UserResponseDto {
     private String firstName;
     private String lastName;
     private String shippingAddress;
+    private Set<Role> roles;
```
Comment (Elena-Bruyako) [inline src/main/java/shop/dto/responce/UserResponseDto.java]:
```suggestion
```

File: src/main/java/shop/entity/User.java [hunk 9]
```diff
@@ -26,4 +40,45 @@ public class User {
     @Column(nullable = false)
     private String lastName;
     private String shippingAddress;
+    @ManyToMany()
```
Comment (Elena-Bruyako) [inline src/main/java/shop/entity/User.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/shop/entity/User.java [hunk 10]
```diff
@@ -26,4 +40,45 @@ public class User {
...
+    )
+    private Set<Role> roles = new HashSet<>();
+    @Column(nullable = false)
+    private boolean isDeleted = false;
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return roles.stream()
+                .map(role -> new SimpleGrantedAuthority("ROLE_" + role.getRole()))
+                .collect(Collectors.toSet());
```
Comment (Elena-Bruyako) [inline src/main/java/shop/entity/User.java]:
```suggestion
        return roles;
```

File: src/main/java/shop/mapper/UserMapper.java [hunk 11]
```diff
@@ -1,13 +1,15 @@
...
 import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
 import shop.config.MapperConfig;
 import shop.dto.request.UserRegistrationRequestDto;
 import shop.dto.responce.UserResponseDto;
 import shop.entity.User;
 
 @Mapper(config = MapperConfig.class)
 public interface UserMapper {
+    @Mapping(target = "roles", source = "roles")
```
Comment (Elena-Bruyako) [inline src/main/java/shop/mapper/UserMapper.java]:
```suggestion
```

File: src/main/java/shop/service/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -23,6 +25,7 @@ public UserResponseDto register(UserRegistrationRequestDto registerUserRequestDt
             throw new RegistrationException("Can't register user with given email");
         }
         User user = userMapper.toUser(registerUserRequestDto);
+        user.setPassword(passwordEncoder.encode(registerUserRequestDto.getPassword()));
```
Comment (Elena-Bruyako) [inline src/main/java/shop/service/impl/UserServiceImpl.java:33]: set ROLE USER
```suggestion
        user.setPassword(passwordEncoder.encode(registerUserRequestDto.getPassword()));
        user.setRole(Set.of(....
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/shop/repository/role/RoleRepository.java [hunk 13]
```diff
@@ -0,0 +1,8 @@
+package shop.repository.role;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+import shop.entity.Role;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
+    Role findByRole(Role.RoleName role);
```
Comment (ahoienko) [inline src/main/java/shop/repository/role/RoleRepository.java]:
```suggestion
    Optional<Role> findByRole(Role.RoleName role);
```

**Review by ahoienko — CHANGES_REQUESTED** [review]
very good, left minor comment

File: src/main/java/shop/repository/role/RoleRepository.java [hunk 14]
```diff
@@ -0,0 +1,9 @@
+package shop.repository.role;
+
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import shop.entity.Role;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
+    Optional<Role> findByRole(Role.RoleName role);
```
Comment (JJJazl) [inline src/main/java/shop/repository/role/RoleRepository.java:8]:
```suggestion
    Optional<Role> findByRoleName(Role.RoleName roleName);
```

File: src/main/java/shop/service/impl/UserServiceImpl.java [hunk 15]
```diff
@@ -23,6 +30,9 @@ public UserResponseDto register(UserRegistrationRequestDto registerUserRequestDt
...
+        user.setRoles(Set.of(roleRepository.findByRole(Role.RoleName.ROLE_USER)
+                .orElseThrow(() -> new EntityNotFoundException("Can`t find role"))));
```
Comment (JJJazl) [inline src/main/java/shop/service/impl/UserServiceImpl.java:35]: Let's make the code more explicit and the exception messages more informative
```suggestion
RoleName userRoleName = Role.RoleName.ROLE_USER
        user.setRoles(Set.of(roleRepository.findByRoleName(userRoleName)
                .orElseThrow(() -> new EntityNotFoundException("Can`t find role: " + userRoleName.name()))));
```

**Review by JJJazl — APPROVED** [review]

**Pull request merged**